src/libsyntax/parse/parser.rs

```diff
@@ -2603,13 +2601,6 @@ impl<'a> Parser<'a> {
     pub fn parse_more_binops(&mut self, lhs: P<Expr>, min_prec: usize) -> P<Expr> {
         if self.expr_is_complete(&*lhs) { return lhs; }
 
-        // Prevent dynamic borrow errors later on by limiting the
-        // scope of the borrows.
-        if self.token == token::BinOp(token::Or) &&
-            self.restrictions.contains(RESTRICTION_NO_BAR_OP) {
-            return lhs;
-        }
```

Comment:
Could you explain the reasoning behind why it is safe to remove this check please?